client/src/index.js

```diff
@@ -1,13 +1,16 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
+import { BrowserRouter as Router } from 'react-router-dom';
 
 import App from './App';
 import GlobalStyles from './GlobalStyles';
 
 ReactDOM.render(
   <React.StrictMode>
-    <GlobalStyles />
-    <App />
+    <Router>
+      <GlobalStyles />
+      <App />
+    </Router>
```

Comment:
Maybe place your `<GlobalStyles />` and `<Router>` inside the `<App />`

Reply:
Can you explain this? We learned and the documention tells the same, that the `<App/>` has to be in the `<Router/>` to work. Just like the code shows?! Or is there something i missed about it?

Reply:
If I'm not wrong, we learned to keep the index.js clean. I like it like this because then your `<App />` is kind of a single source of truth. You get a clean structure of your app where you can see immediately which components are included. I think it would be also better for testing. But maybe it's just a matter of taste ... 😉

@lmachens any suggestions?

Reply:
Hi :), sorry for this long answer. Please keep in mind, that this is a complex topic with different opinions. It's not wrong to add it to `index.js`. Let me describe, why I personally think there is a better solution. 

The `<App />` requires `<Router>` because of routes and links. If you don't have a `Router`, the App won't work.
Thus, you always have to add a Router when you call `App`. This makes it more complex if you like to test or render it in isolation. Good examples are Storybook, Jest or SSR where you have to add a router and the GlobalStyles too. If you have more providers for themes, caching, storage or requests, you might add them to the `index.js` too. In this case, you have even more to add in the entry points.

The reason, why `Router` might be extracted from the App is, that Jest requires a different Router (MemoryRouter) or likes to set an initial route. 
```jsx
test('full app rendering/navigating', () => {
  const history = createMemoryHistory()
  render(
    <Router history={history}>
      <GlobalStyles />
      <App />
    </Router>
  )
```
See https://testing-library.com/docs/example-react-router/
You could say, that you use the App in a different context (in a memory router instead of a browser router). 
A similar example, if you like to test the app with a different theme:

```jsx
test('full app rendering in dark mode', () => {
  render(
    <ThemeProvider theme="dark">
      <GlobalStyles />
      <App />
    </ThemeProvider >
  )
```

The problem is, that you have to add the other providers like `Router` too. 

```jsx
test('full app rendering in dark mode', () => {
  const history = createMemoryHistory()
  render(
    <Router history={history}>
      <ThemeProvider theme="dark">
        <GlobalStyles />
        <App />
      </ThemeProvider >
    </Router>
  )
```

This is the reason, why you might add a helper component or function (https://testing-library.com/docs/example-react-
router/#reducing-boilerplate). But it's getting complex :-). 

I think, the app should work without additional providers and include everything it needs for rendering in production.
This is possible if you add default providers like the BrowserRouter and ThemeProvider to your `App.js`.

```jsx
const App = () => (
  <Router history={history}>
      <ThemeProvider theme="dark">
        <GlobalStyles />
        <Switch>
          ....
        </Switch>
      </ThemeProvider >
    </Router>
);
```

Now, you need to know about mocking https://jestjs.io/docs/manual-mocks. It allows you to change the default behavior of your code like creating fake responses to fetches. If you don't do this, you have to create a Provider for fake requests and a Provider for real requests.

Example:
```jsx
jest.mock('react-router-dom', () => ({
  BrowserRouter: ({ children }) => <div>{children}</div>,
}));

test('full app rendering/navigating', () => {
  const history = createMemoryHistory()
  render(
    <Router history={history}>
      <App />
    </Router>
  )
```

This prevents the default behavior of `BrowserRouter` and don't require adding other providers or the GlobalStyles.

The `index.js` is the entry point where you load your applications. `App` is one application, the service worker is a different application. I think this should be as clean as possible without other context providers.

Reply:
Thank you, Leon, for the detailed explanation. I will leave the Router in the index.js for now but it's still good to know that this could cause problems in some cases.